### DISCUSSION TITLE
Blazor server side empty objects in dataportal

### DISCUSSION THREAD

**Discussion by Mikael-Pothier** [MarimerLLC/csla#2216] [Questions] [answered]
Hello,

I am new to Blazor and I am trying to insert an object using the dataportal. I create the object using this code

```c#
                var newCourtier = Courtier.NewCourtier();

                newCourtier.NoCourtier = courtierToAdd.NoCourtier;
                newCourtier.Adresse1 = courtierToAdd.Adresse1;
                newCourtier.Province = courtierToAdd.Province;
                newCourtier.CodePostal = courtierToAdd.CodePostal;
                newCourtier.Email = courtierToAdd.Email;

                return Task.Run(() => newCourtier.SaveAsync()).Result;
```

When it goes in SaveAsync, it seems to create a new object because when it goes into DataPortal_Insert() my object is empty.

Thank you for your help.

**Comment by rockfordlhotka** [comment, marked as answer]
I'd expect the code to be more like this:

```c#
  var newCourtier = await Courtier.NewCourtierAsync();

  newCourtier.NoCourtier = courtierToAdd.NoCourtier;
  newCourtier.Adresse1 = courtierToAdd.Adresse1;
  newCourtier.Province = courtierToAdd.Province;
  newCourtier.CodePostal = courtierToAdd.CodePostal;
  newCourtier.Email = courtierToAdd.Email;

  return await newCourtier.SaveAsync();
```

Client-side Blazor is single threaded (like browsers, JavaScript, and WebAssembly), so I have no idea what happens if you try to do Task.Run stuff.

**Reply by Mikael-Pothier** [reply]
Thank you for your fast answer. I tried this and in the DataPortal I still have an empty object.

One thing is I'm not using Client-side blazor, but server side blazor. My razor page which does the insert call a service and my services call the saveasync.

I did a test to show you what happened.
Here is what I have in my noCourtier when it is in my services 
![image](https://user-images.githubusercontent.com/8630076/115264929-c1e6c680-a104-11eb-8dd8-7c30263ca420.png)

and here is what I have in the dataportal when I tried to save
![image](https://user-images.githubusercontent.com/8630076/115265115-e93d9380-a104-11eb-8423-2740474009b3.png)

**Reply by rockfordlhotka** [reply]
If you are server-side Blazor then _maybe_ it will work with Task.Run and .Result - those aren't scenarios I've tested or would ever recommend.

But let's step back and consider your `Courtier` class. It is using managed backing fields, inherits from BusinessBase, etc? In other words, we have confidence that this business class works outside Blazor?

**Reply by Mikael-Pothier** [reply]
Yes the class works outside of blazor we have a test program and it tests the CRUD operation and all of them are working in the `courtier` class. 

For the courtier class the code is made from a code generator(codesmith) and the class inherits from BusinessBase and all the property looks like this one

```c#
private static readonly PropertyInfo<System.Int16> _noCourtierProperty = RegisterProperty<System.Int16>(p => p.NoCourtier, "No Courtier", RelationshipTypes.PrivateField);`
        
private System.Int16 _noCourtier = _noCourtierProperty.DefaultValue;`

        [System.ComponentModel.DataObjectField(true, false)]
        public System.Int16 NoCourtier
        {
            get { return GetProperty(_noCourtierProperty, _noCourtier); }
            set { SetProperty(_noCourtierProperty, ref _noCourtier, value); }
        }
```

**Reply by rockfordlhotka** [reply]
Is this with a local or remote data portal? Also, can you share your CSLA config from `Startup.cs`?

**Reply by rockfordlhotka** [reply]
Ahh, yes, sorry, haven't used private backing fields for an extremely long time, and some extra work is required when using them with `MobileFormatter`.

Your class needs to override `OnGetState` and `OnSetState` to get/set the private backing field into the serialization stream. Something like this in your business class:

```c#
    protected override void OnGetState(SerializationInfo info, StateMode mode)
    {
      info.AddValue(_noCourtierProperty.Name, _noCourtier);
      base.OnGetState(info, mode);
    }

    protected override void OnSetState(SerializationInfo info, StateMode mode)
    {
      _noCourtier = info.GetValue<Int16>(_noCourtierProperty.Name);
      base.OnSetState(info, mode);
    }
```

This is because CSLA (especially in Blazor) uses `MobileFormatter` by default. Microsoft is phasing out `BinaryFormatter`, which is what _used to be the default_, and so CSLA is following Microsoft's lead.

So your options are:

1. Stop using private backing fields; CSLA handles managed backing fields automatically
2. Modify your code-gen templates to override the get/set methods
3. If you have a custom base class between (for example) BusinessBase and your business class, you can add a bit of reflection in reusable get/set methods to auto-get/set all private backing fields (this was discussed in a different thread within the last few weeks)

**Reply by Mikael-Pothier** [reply]
I removed the managed backing fields and now the save work.

I have another question

since we are using Blazor we used `appSetting.json` instead of `app.config` and we cannot use the `ConfigurationManager` for the connectionstring and since we are not using entityframework we couldn't make it work with `services.AddDbContext`. We found a way to do it, but I'm not sure if it's the best practice. 

What we did was to set in the `Configure(IApplicationBuilder app, IWebHostEnvironment env)` of `startup.cs` we set the connectionstring in or databaseHelperclass (with the one we set in appsettings.json) like this
`Library.ADOHelper.ConnectionString = Configuration.GetConnectionString("TestPrimacoConnectionString");`
Do you know a better way or this way is valid?

Thank you for your help!!!

**Reply by rockfordlhotka** [reply]
I'm glad this is working for you. I think it best if you post your new question in its own thread.
